### PR TITLE
Code verifier & code challenge generator for working with PKCE

### PR DESCRIPTION
**Pull request by MrWoWander**
To work with PKCE, it is more convenient to have a code verifier and code challenge generator in the framework. 
Code challenge generation works according to the "SHA256" standard

### CODE REVIEW

**Comment by phimage** [comment]
thanks, every body could copy you code if I do not merge it 
(similar I think than the code in auth0 doc)

The reason that I do not merge immediately is the grip on `CommonCrypto`
we could add maybe macro if canImport

even the lib I think is not linux compatible this add a new issue
for linux and swiftpm there is 
https://github.com/apple/swift-crypto
I do it one time here https://github.com/IBDecodable/IBLinter/blob/master/Package.swift#L59 , but maybe there is more clean way now because SwiftPM had some evolution

**Comment by MrWoWander** [comment]
> thanks, every body could copy you code if I do not merge it
> (similar I think than the code in auth0 doc)
> 
> The reason that I do not merge immediately is the grip on `CommonCrypto`
> we could add maybe macro if canImport
> 
> even the lib I think is not linux compatible this add a new issue
> for linux and swiftpm there is
> https://github.com/apple/swift-crypto
> I do it one time here https://github.com/IBDecodable/IBLinter/blob/master/Package.swift#L59 , but maybe there is more clean way now because SwiftPM had some evolution

Then I have a suggestion to add a new library to SPM: Perfect-Crypto. She herself knows how to generate SHA256
https://github.com/PerfectlySoft/Perfect-Crypto#digest

Well, or I can try to get confused, and write my translation in SHA256 by analogy, as in this library =)

**Comment by phimage** [comment]
same here, I do not want to get all crypto function for one need (common crypto is at least embedded in iOS)

If a person have a have with common crypto, crypto swift, or perfect-crypto already in its app I do not want add a new lib

**Comment by MrWoWander** [comment]
Okay. I have limited the given methods from Linux.

> same here, I do not want to get all crypto function for one need (common crypto is at least embedded in iOS)
> 
> If a person have a have with common crypto, crypto swift, or perfect-crypto already in its app I do not want add a new lib
> 
> (I have same issue with RSA)

**Comment by MrWoWander** [comment]
I suggest this as a temporary solution, as several people have already encountered ```PKCE```

**Comment by phatblat** [comment]
This could be made portable to Linux if you use Swift [Crypto](https://swift.org/blog/crypto/) instead of CommonCrypto

**Comment by MrWoWander** [comment]
> This could be made portable to Linux if you use Swift [Crypto](https://swift.org/blog/crypto/) instead of CommonCrypto

On your advice, it seems like it turned out to implement support for SHA256

**Comment by phimage** [comment]
I think it's ok and could be merged

I am just wondering if we keep as global function or put at static extension of OAuthSwift class

PS: Code do not compile on linux yet, a lot of `if os(target)` , import of uikit and appkit
things like importing `FoundationNetworking` I think also

I have made a test here https://github.com/OAuthSwift/OAuthSwift/runs/2465840429 or here https://github.com/OAuthSwift/OAuthSwift/runs/2465733255

File: README.md [hunk 1]
```diff
@@ -179,6 +179,8 @@ let handle = oauthswift.authorize(
 ```
 
 ### Authorize with OAuth2.0 and proof key flow (PKCE)
+
+*Methods ```generateCodeVerifier``` and ```generateCodeChallenge``` are ***not compatible*** with ```Linux```*
```
Comment (phimage) [inline README.md:183]: Could be removed now?

File: Sources/Utils.swift [hunk 2]
```diff
@@ -21,3 +27,42 @@ public func generateState(withLength len: Int) -> String {
     }
     return randomString
 }
+
```
Comment (phimage) [inline Sources/Utils.swift:30]: only one blank line, I think SwiftLint will warn about it

**Review by phimage — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (MrWoWander) [reply README.md]: Done

File: Sources/Utils.swift (hunk 2 above)
Comment (MrWoWander) [reply Sources/Utils.swift:30]: Done

**Comment by phimage** [comment]
I merge it, I will fix other SwiftLint warnings using `swiftlint autocorrect`

**Pull request merged**